kube-runtime/src/reflector/object_ref.rs

```diff
@@ -1,40 +1,25 @@
 use derivative::Derivative;
 use k8s_openapi::{api::core::v1::ObjectReference, apimachinery::pkg::apis::meta::v1::OwnerReference};
-use kube_client::{
-    api::{DynamicObject, Resource},
-    core::ObjectMeta,
-    ResourceExt,
-};
-use std::{
-    fmt::{Debug, Display},
-    hash::Hash,
-};
+use kube_client::core::{ObjectMeta, TypeInfo, TypeMeta};
+use std::fmt::{Debug, Display};
+use thiserror::Error;
 
 #[derive(Derivative)]
-#[derivative(
-    Debug(bound = "K::DynamicType: Debug"),
-    PartialEq(bound = "K::DynamicType: PartialEq"),
-    Eq(bound = "K::DynamicType: Eq"),
-    Hash(bound = "K::DynamicType: Hash"),
-    Clone(bound = "K::DynamicType: Clone")
-)]
-/// A typed and namedspaced (if relevant) reference to a Kubernetes object
-///
-/// `K` may be either the object type or `DynamicObject`, in which case the
-/// type is stored at runtime. Erased `ObjectRef`s pointing to different types
-/// are still considered different.
+#[derivative(Debug, PartialEq, Eq, Hash, Clone)]
+
+/// A named, typed and namedspaced reference to a Kubernetes object
 ///
 /// ```
+/// use kube_client::core::{Resource};
+/// use k8s_openapi::api::{apps::v1::Deployment, core::v1::Pod};
 /// use kube_runtime::reflector::ObjectRef;
-/// use k8s_openapi::api::core::v1::{ConfigMap, Secret};
 /// assert_ne!(
-///     ObjectRef::<ConfigMap>::new("a").erase(),
-///     ObjectRef::<Secret>::new("a").erase(),
+///     ObjectRef::new("a").within("ns").with_types(&<Deployment as Resource>::typemeta().unwrap()),
+///     ObjectRef::new("a").within("ns").with_types(&<Pod as Resource>::typemeta().unwrap()),
 /// );
 /// ```
 #[non_exhaustive]
-pub struct ObjectRef<K: Resource> {
-    pub dyntype: K::DynamicType,
+pub struct ObjectRef {
```

Comment:
Not a fan of losing the type information here. Especially how this effectively makes it possible (and even the default!) to make a completely untyped `ObjectRef`, which is ~never valid.

Treating `namespace: None` as a wildcard has been problematic enough so far IMO. :/

Reply:
If we try to lift the constraint of `TypeInfo` up to be a core requirement of the `ObjectRef`, is that more acceptable? I think that might require some rework in some test edge cases, but it would create a more palatable default (the `TypeInfo` trait can in theory return `None` types, but practically it never does for watch cases).